### DISCUSSION TITLE
Instructions required, how to install on windows without chocolatey

### DISCUSSION THREAD

**Discussion by aisbergde** [schemacrawler/SchemaCrawler#449] [Ideas]
I used chocolatey to install SchemaCrawler on Windows:
https://chocolatey.org/packages/schemacrawler
but the latest available package is still 16.4.1, the package 16.4.2 is still not approved: https://chocolatey.org/packages/schemacrawler/16.14.2

I've been waiting for several days now so I can test the new version. I tried
`choco upgrade schemacrawler --version 16.4.2`
but it doesn't work.

Then my idea was just to download the latest version from here: https://github.com/schemacrawler/SchemaCrawler/releases
But I have no idea how to use it to "install" on Windows. There is no exe or anything that would simply work under Windows. Maybe there is also something, but then there is no instruction.

https://github.com/schemacrawler/SchemaCrawler-Installers doesn't help me to understand what is required to install.

The chocolatey setup creates a **Windows exe file** and this works fine on Windows:
- the exe is placed there: "C:\ProgramData\chocolatey\bin\schemacrawler.exe"
- C:\ProgramData\chocolatey\bin is included in the path
- schemacrawler.exe can "just be used"
- an environment variable is also created and maintained (not sure if this is required):
  `SC_HOME=C:\ProgramData\chocolatey\lib\schemacrawler\schemacrawler-16.14.1-distribution`

Please provide some instruction how to install SchemaCrawler on a Windows system without chocolatey. Please write this into the documentation. The instruction should be so understandable that I can easily give Windows users a link to this step-by-step instruction where it is described how to install SC on Windows - without chocolatey

**Comment by adriens** [comment]
Working on it see https://github.com/adriens/chocolatey-schemacrawler/issues/56

**Comment by adriens** [comment]
Currently trying a from scratch install.. with `choco install schemacrawler --version=16.14.2`

**Comment by adriens** [comment]
Got this : 

```
Installed:
 - chocolatey-core.extension v1.3.5.1
 - javaruntime v8.0.231
 - schemacrawler v16.14.2
 - graphviz v2.47.0

Failures
 - jre8 (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\jre8\tools\chocolateyInstall.ps1'.
 See log for details.

C:\WINDOWS\system32>schemacrawler -version
SchemaCrawler 16.14.2

Error: Missing required options: '--command=<command>', '--info-level=<infolevel>'

Run SchemaCrawler again with just the
-h
option for help

Or, run SchemaCrawler again with an additional
--log-level=CONFIG
option for details on the error

Additional resources are on the SchemaCrawler website,
https://www.schemacrawler.com
and a live, interactive tutorial is on
https://katacoda.com/schemacrawler
```

Currently you could : 

```
choco uninstall schemacrawler
choco upgrade schemacrawler --version=16.14.2
```

I will try to reproduce the upgrade issue on https://github.com/adriens/chocolatey-schemacrawler/issues/56

**Comment by adriens** [comment]
Oh, so don't want chocolatey .. did not see.
So I'm passing my turn.

**Reply by aisbergde** [reply]
@adriens You misinterpreted this. I prefer to use choco whenever this is possible. But in some environments this is not an option. That's why I need additional ways to install in the case, when choco is not an option. or to recommend SchemaCrawler in General. And I want to thank you for presenting SchemaCrawler on choco!

**Comment by adriens** [comment]
@aisbergde  : on which systems woud you like to get an installer please ?

**Reply by aisbergde** [reply]
@adriens only Windows systems. Normally I am working in projects where I can't use chocolatey. And I need solutions without admin rights required. I also want to recommend SchemaCrawler in General, but this should work without admin right and without installing content in some "special folders".

Often the target system is some Citrix environment, it is managed by the IT department, and they don't synchronize all folders. That's why the best would be any solution which can be used without any admin rights and to be placed in "normal" folders. Zip archives are fine, but how to add the windows specific context to be able to "just execute"?

In other words, a choco-like solution, to be used in this context.

**Comment by sualeh** [comment]
@aisbergde - there are several ways to use SchemaCrawler on Windows, and you can choose the one that works for you. You will need Java installed beforehand.
1. Chocolatey: The Chocolatey installer is maintained by @adriens and he keeps it up to the latest version. There are ways to use Chocolatey without admin rights, and you will have to research these on the Chocolatey website.
2. winget: [winget, or Windows Package Manager](https://docs.microsoft.com/en-us/windows/package-manager/) is a Microsoft installation tool, and SchemaCrawler is available for installation using winget. However, winget is still in preview mode. The SchemaCrawler package on winget is maintained by me.
3. SchemaCrawler installers: You can get an MSI file to install the SchemaCrawler Shell from [SchemaCrawler-Installers](https://github.com/schemacrawler/SchemaCrawler-Installers/releases), but will need admin privileges and will have to work through the dialogs to install. The SchemaCrawler Installers project is maintained by me.
4. SchemaCrawler releases: You can get a zip file with SchemaCrawler from [SchemaCrawler releases](https://github.com/schemacrawler/SchemaCrawler/releases). If you unzip the file, you will find a "_schemacrawler" directory, which is all you need. That has a "cmd" file to run SchemaCrawler. No admin privileges are required. You will have to figure out how to put this on your system PATH. SchemaCrawler release are maintained by me.

**Comment by aisbergde** [comment]
(1) and (3) are working fine. Best with Admin rights.

It looks like (4) should work fine without admin rights. I just tested it.

I did not know that I just can use it as it is. When I executed it using a double click then I got this strange warnings, which can't be solved on systems without admin rights. But when I first open a command line and then call the `schemacrawler.cmd`, then it works fine.

(2) winget - I never used this. I will test this. Maybe Microsoft will push it, and then it will be a good idea too. It looks like I can install the "App-Installer" on all my PC's even remotely. This is interesting to keep all PC in my family up to date with supported software.

=> anyway, there is minimum one way to get this installed without admin rights, but just using copy and paste!

**Reply by sualeh** [reply]
5. You can run SchemaCrawler from Docker on Windows. No need to install anything but Docker, not even Java. The SchemaCrawler Docker image is maintained by me.

**Reply by aisbergde** [reply]
I have never used Docker, but have read so much about it. Maybe this could be even easier if it was clear how easy it is to use on Windows. At some point I had tried something with Docker, it didn't work, and since then Docker has the stamp: "Complicated and doesn't work" with me. Maybe it's worth a second try.

**Reply by sualeh** [reply]
Docker on Windows has come a long way, and it is time to try it again. If you use it, you will not need to install Java, Graphviz or Chocolatey. You can run a command like `docker run --mount type=bind,source="$(pwd)",target=/home/schcrwlr --rm -it schemacrawler/schemacrawler /opt/schemacrawler/schemacrawler.sh --url=... --user=... --password=... --database=... --info-level=standard --command=schema --output-file=output.png`

**Reply by aisbergde** [reply]
Maybe it would be really easy to use Docker if I knew how to start. Docker have installed. `docker/getting started` I also get to work.

But I can't do much with your tutorial (https://github.com/schemacrawler/SchemaCrawler-Docker) because the syntax is not for Windows: So first I would have to know how to write the commands written there so that they work under Windows.
Where this should be executed?

```
docker run \
-v $(pwd):/share \
--rm -i -t \
--entrypoint=/opt/schemacrawler/schemacrawler-shell.sh \
schemacrawler/schemacrawler
```

It is not working in the command line, it is not working in Powershell. I think this is for linux, unix or some kind of these operating systems. I don't know unix or linux syntax.

I also tried this, without success.

```
D:\SchemaCrawler>docker run --mount type=bind,source="$(pwd)",target=/home/schcrwlr --rm -it schemacrawler/schemacrawler /opt/schemacrawler/schemacrawler.sh --url=jdbc:sqlserver://localhost;databaseName=AdventureWorks2017; --schemas=AdventureWorks2017\..* --info-level=standard --command=list --user=documenter --password=documenter --info-level=standard --command=schema --output-file=output.png
docker: Error response from daemon: invalid mount config for type "bind": invalid mount path: '$(pwd)' mount path must be absolute.
See 'docker run --help'.
```

I'm afraid that it will now take me hours or days to understand how to get SchemaCrawler (or whatever app) to work with Docker.

**Reply by aisbergde** [reply]
https://www.schemacrawler.com/docker-image.html

links to

https://hub.docker.com/r/schemacrawler/schemacrawler/

containing a pull command:

```
docker pull schemacrawler/schemacrawler
```

the pull works

but now? I could try this (like getting started). But I think, this is not the right way.

```
docker run -dp 80:80 schemacrawler/schemacrawler
```

**Comment by adriens** [comment]
Maybe should you give a try to https://www.katacoda.com/schemacrawler

**Reply by aisbergde** [reply]
I used all these katacoda examples to learn SC. And the question is not, how to use these examples, but how to run SC in Docker with my databases. On my system I can use it with the choco install, it works fine. But the question is: how to use it on other systems. Docker would be fine, if the documentation would explain, what to do on Windows.

**Reply by sualeh** [reply]
@aisbergde - try `docker run --mount type=bind,source="$(pwd)",target=/home/schcrwlr --rm -it schemacrawler/schemacrawler /opt/schemacrawler/schemacrawler.sh --url=... --user=... --password=... --database=... --schemas=... --info-level=minimum --command=list --output-file=output.png`

**Reply by aisbergde** [reply]
as I wrote before:

>docker: Error response from daemon: invalid mount config for type "bind": invalid mount path: '$(pwd)' mount path must be absolute.

**Reply by sualeh** [reply]
@aisbergde - I use Docker Compose to use SchemaCrawler, it makes the volume mounts easy, even on Windows. The Docker Compose files are at [schemacrawler/SchemaCrawler/schemacrawler-docker-compose](https://github.com/schemacrawler/SchemaCrawler/tree/master/schemacrawler-docker-compose).

**Comment by adrianboangiu** [comment]
I have also pass through the same experience like aisbergde. Meanwhile katacoda has died.
I succeed to use SC with docker on Windows and I will put here the command to run SC for others who might need it:

So, after installing and starting Docker Desktop, you need to pull SC by running the following command in a command prompt:
`docker pull schemacrawler/schemacrawler` 

To run SC you need to execute the following command in a command prompt:
`docker run -v %cd%:/home/schcrwlr/share --name schemacrawler --rm -i -t --entrypoint=/bin/bash schemacrawler/schemacrawler`

You can launch the SC shell afterwards with:
`schemacrawler --shell`

Thats all.
